client/components/sticky-panel/index.jsx

```diff
@@ -16,31 +16,151 @@ import afterLayoutFlush from 'lib/after-layout-flush';
 	};
+}
+
+function renderStickyPanel( props, state ) {
+	const classes = classNames( 'sticky-panel', props.className, {
+		'is-sticky': state.isSticky,
+	} );
+
+	return (
+		<div className={ classes }>
+			<div className="sticky-panel__content" style={ getBlockStyle( state ) }>
+				{ props.children }
+			</div>
+			<div className="sticky-panel__spacer" style={ { height: state.spacerHeight } } />
+		</div>
+	);
+}
+
+function isWindowTooSmall( minLimit ) {
+	return ( minLimit !== false && minLimit >= window.innerWidth ) || isMobile();
+}
+
+class StickyPanelWithIntersectionObserver extends React.Component {
+	static displayName = 'StickyPanel';
 
-	static defaultProps = {
-		minLimit: false,
+	static propTypes = commonPropTypes;
+	static defaultProps = commonDefaultProps;
+
+	state = {
+		isSticky: false,
+		spacerHeight: 0,
+		blockWidth: 0,
 	};
 
+	onIntersection = afterLayoutFlush( entries => {
+		if ( ! entries || ! entries[ 0 ] ) {
+			return;
+		}
+		const { intersectionRatio, rootBounds, boundingClientRect } = entries[ 0 ];
+		const isSticky = intersectionRatio < 1 && rootBounds.bottom > boundingClientRect.bottom;
+		this.updateStickyState( isSticky );
+	} );
+
+	throttleOnResize = throttle(
+		() => this.setState( prevState => getDimensions( this, prevState.isSticky ) ),
+		RESIZE_RATE_IN_MS
+	);
+
+	componentDidMount() {
+		window.addEventListener( 'resize', this.throttleOnResize );
+
+		this.deferredMount = defer( () => {
+			this.observer = new IntersectionObserver( this.onIntersection, {
+				threshold: [ 0, 1 ],
+				rootMargin: `-${ calculateOffset() }px 0px 0px 0px`,
+			} );
+			this.observer.observe( ReactDom.findDOMNode( this ) );
+		} );
+	}
+
+	componentWillUnmount() {
+		if ( this.observer ) {
+			this.observer.disconnect();
+		}
+		this.onIntersection.cancel();
+		clearTimeout( this.deferredMount );
+		window.removeEventListener( 'resize', this.throttleOnResize );
+	}
+
+	updateStickyState( isSticky ) {
+		if ( isWindowTooSmall( this.props.minLimit ) ) {
+			return this.setState( { isSticky: false } );
+		}
+
+		if ( isSticky !== this.state.isSticky ) {
+			this.setState( {
+				isSticky,
+				...getDimensions( this, isSticky ),
+			} );
+		}
+	}
+
+	render() {
+		return renderStickyPanel( this.props, this.state );
+	}
+}
+
+class StickyPanelWithScrollEvent extends React.Component {
+	static displayName = 'StickyPanel';
+
+	static propTypes = commonPropTypes;
+	static defaultProps = commonDefaultProps;
+
 	state = {
 		isSticky: false,
 		spacerHeight: 0,
 		blockWidth: 0,
 	};
 
+	onWindowScroll = afterLayoutFlush( () => {
+		// Determine vertical threshold from rendered element's offset relative the document
+		const threshold = ReactDom.findDOMNode( this ).getBoundingClientRect().top;
+		const isSticky = threshold < calculateOffset();
+		this.updateStickyState( isSticky );
+	} );
+
+	throttleOnResize = throttle(
+		() => this.setState( prevState => getDimensions( this, prevState.isSticky ) ),
```

Comment:
I'm not sure if it's really important, but we could avoid calling `setState` too often when the dimensions don't change. Especially in the case of `isSticky === false`, they are guaranteed to be `0` all the time.

Reply:
I added a `getDimensionUpdates` method that only returns an object when there are in fact changes, and `null` otherwise. Returning a `null` from the function passed to `setState` cancels the update.